File: Java/3_Longest_Substring_Without_Repeating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        //left and right store window values
        int l = 0, r = 0, sol = 0;
        int length = s.length();
        //set stores values weve seen
        HashSet<Character> set = new HashSet<>();
        while(r < length) {
            //if set does contain char then we expand window
            //and remove start of window
            if(set.contains(s.charAt(r))) {
                set.remove(s.charAt(l));
                l++;
            } else {
                //otherwise we add to set
                set.add(s.charAt(r));
                //check if we have our new longest substring
                sol = Math.max(sol, r - l + 1);
                //and keep going
                r++;
            }
        }
        //return highest value we got
        return sol;
    }
}

File: Python/3_Longest_Substring_Without_Repeating_Characters.py
#Write up: The code works by having a sliding window that checks
# if the current window contains a duplicate,
#if it does not, it goes further to find the largest possible window
# and checks if we have found a new longest substring
# if it does contain a duplicate, it cuts the substring off by one from
#the start and then checks again. The loop will end when the right pointer
# has reached the end of the string and then returns the maximum length
# we found. The time complexity of the algorithm is O(n) because each
#charcter is looked at at most twice, which is 2n times in total
# The space complexity is O(n) because the HashMap stores at most n
# characters if the string has no duplicates
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
    #initalizes left and right pointers and variable that tracks solution
        l,r,sol = 0,0,0
    #initalizes set that holds values present in current window
        curr_elems = set()
    #loop runs while we have not reached the end
        while r < len(s):
        #checks case where a duplicate element is found
            if s[r] in curr_elems:
            #cuts off front of window and tries again
                curr_elems.remove(s[l])
                l += 1
            else:
            #checks case where a duplicate is NOT found
            #adds new character to set of present characters
                curr_elems.add(s[r])
            #checks if new longest substring found
                sol = max(sol, r - l + 1)
            #increases window
                r += 1

    #returns solution
        return sol

